plugins/modem-manager/fu-mbim-qdu-updater.h

```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright (C) 2021 Jarvis Jiang <jarvis.w.jiang@gmail.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#ifndef __FU_MBIM_QDU_UPDATER_H
+#define __FU_MBIM_QDU_UPDATER_H
+
+#include <libmbim-glib.h>
+#include "fu-mm-device.h"
+
+#define FU_TYPE_MBIM_QDU_UPDATER (fu_mbim_qdu_updater_get_type ())
+G_DECLARE_FINAL_TYPE (FuMbimQduUpdater, fu_mbim_qdu_updater, FU, MBIM_QDU_UPDATER, GObject)
+
+FuMbimQduUpdater *fu_mbim_qdu_updater_new	(const gchar *mbim_port);
+gboolean fu_mbim_qdu_updater_open	(FuMbimQduUpdater	*self,
+					 GError 		**error);
+GArray *fu_mbim_qdu_updater_write	(FuMbimQduUpdater	*self,
+					 const gchar		*filename,
+					 GBytes 		*blob,
+					 GError 		**error,
+					 FuDevice 		*device);
```

Comment:
If you do make the derivation as I suggested you no longer pass both a `FuDevice` and `FuMbimQduUpdater` both as arguments.

Reply:
I try to pass the FuMmDevice argument instead of FuDevice and FuMbimQduUpdater as you suggested, but get a make error: dereferencing pointer to incomplete type 'FuMmDevice ' {aka 'struct _FuMmDevice '}

Could you tell me how to solve it? Thanks.

Reply:
You'll have to make FuMbimQduUpdater derive from `FuMmDevice`, change it's gobject type in your header and source file.

Reply:
I changed the fu-mbim-qdu-updater.h
GArray *fu_mbim_qdu_updater_write (**FuMmDevice** *dev, const gchar *filename, GBytes *blob, GError **error);

and fu-mbim-qdu-updater.c
GArray *fu_mbim_qdu_updater_write (**FuMmDevice** *dev, const gchar *filename, GBytes *blob, GError **error)
{

        g_autoptr(GMainLoop) mainloop = g_main_loop_new (NULL, FALSE);
        g_autoptr(GArray) digest = fu_mbim_qdu_updater_get_checksum (blob);
        FuMbimQduUpdater *mbim_qdu_updater = dev->mbim_qdu_updater;    // added
        WriteContext ctx = {
		.mainloop = mainloop,
		.mbim_device = mbim_qdu_updater ->mbim_device,    // changed
		.cancellable = g_cancellable_new (),
		.error = NULL,
		.timeout_id = 0,
		.blob = blob,
		.digest = digest,
		.chunks = NULL,
		.chunk_sent = 0,
		.device = dev->parent_instance,    // changed
        };

        fu_mbim_qdu_updater_set_update_session (mbim_qdu_updater->mbim_device, &ctx);    // changed
        g_main_loop_run (mainloop);

        if (ctx.error != NULL) {
		g_propagate_error (error, ctx.error);
		return NULL;
        }

        return g_steal_pointer (&digest);
}

But in the place `FuMbimQduUpdater *mbim_qdu_updater = dev->mbim_qdu_updater;    // added`
get a make error: dereferencing pointer to incomplete type 'FuMmDevice ' {aka 'struct _FuMmDevice '}

Is this change you suggest to do? I don't know how to solve the error.

Reply:
> You'll have to make FuMbimQduUpdater derive from FuMmDevice, change it's gobject type in your header and source file.

Not completely sure this is a good idea; it would force you to create a specific new object type at `fu_plugin_mm_device_add()`, and that is not something we do for other upgrade methods. The rationale for this is that a single modem may have more than 1 "updater" logic, and even more, we may have combinations of them...

In my mind, I think that keeping the FuMmDevice without subclasses would be cleaner; as soon as MM reports a new modem with firmware upgrade capabilities we create a FuMmDevice, and then we'll see later on which are the actual update methods we'll be using (i.e. which updater helper objects we'll be using).

@superm1 not sure what you think about that

Reply:
I think you're right, thanks for giving it consideration.